h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -540,95 +533,90 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
         //    int[] colIndex = new int[]     {0, 2,   0, 3,   0, 1, 2};      //col index for each non-zero
 
         // extract predictors
-        int nz = 0;
+        int nonZeroCount = 0;
         int currentRow = 0;
         int currentCol = 0;
         int rwRow = 0;
 
         for (Integer chunk : chunks) {
             for(long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk+1]; i++) {
                 if (w != null && w.at(i) == 0) continue;
-                int nzstart = nz;
+
+                final int startNonZeroCount = nonZeroCount;
                 // enlarge final data arrays by 2x if needed
 
                 for (int j = 0; j < di._cats; ++j) {
-                    if (!vecs[j].isNA(i)) {
-                        data[currentRow][currentCol] = 1; //one-hot encoding
+                    data[currentRow][currentCol] = 1; //one-hot encoding
+                    if (vecs[j].isNA(i)) {
+                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, Float.NaN);
+                    } else {
                         colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
-                        nz++;
                     }
+                    nonZeroCount++;
                 }
 
                 for (int j = 0; j < di._nums; ++j) {
                     float val = (float) vecs[di._cats + j].at(i);
-                    if (!Float.isNaN(val) && val != 0) {
+                    if (val != 0) {
                         data[currentRow][currentCol] = val;
                         colIndex[currentRow][currentCol++] = di._catOffsets[di._catOffsets.length - 1] + j;
-                        nz++;
+                        nonZeroCount++;
                     }
                 }
-                if (nz == nzstart) {
+                if (nonZeroCount == startNonZeroCount) {
                     // for the corner case where there are no categorical values, and all numerical values are 0, we need to
                     // assign a 0 value to any one column to have a consistent number of rows between the predictors and the special vecs (weight/response/etc.)
                     data[currentRow][currentCol] = 0;
                     colIndex[currentRow][currentCol++] = 0;
-                    nz++;
+                    nonZeroCount++;
```

Comment:
Not sure I get this part.  If the whole chunk is 0, then, it should be left alone and we save memory by not recording it.  There must be some other way to keep track of number of rows between the predictors and weight vectors.  I think python may not have done this....

Reply:
This is not about whole chunk, but about whole row. If a whole row is zero-only, then there must be a record somewhere that there is an empty row (for proper reconstruction of the full matrix, if nothing else). Scipy duplicate row index, @mdymczyk 's API uses one more zero and also puts a row index into row headers. This API was there from the beginning and I don't want to break it.

It seems to be interpreted well and other tests only confirm it. In the code above, you can see however that the C code only flattens the two-dim array and passes the flattened array to XGBoost.